src/main/java/org/apache/maven/shared/artifact/filter/PatternIncludesArtifactFilter.java

```diff
@@ -255,540 +183,319 @@ public void reportMissedCriteria( final Logger logger )
         {
-            return VersionRange.createFromVersionSpec( range ).containsVersion( new DefaultArtifactVersion( version ) );
-        }
-        catch ( final InvalidVersionSpecificationException e )
-        {
-            return false;
+            String str = stokens[i];
+            tokens[i] = str != null && !str.isEmpty() ? str : ANY;
         }
+        return tokens;
     }
 
-    static Pattern compile( String pattern )
+    private static Pattern compile( String pattern )
     {
         if ( pattern.startsWith( "!" ) )
         {
             return new NegativePattern( pattern, compile( pattern.substring( 1 ) ) );
         }
         else
         {
-            char[][] stokens = tokenizeAndSplit( pattern );
-            char[][] tokens = new char[ stokens.length ][];
-            for ( int i = 0; i < stokens.length; i++ )
-            {
-                tokens[i] = anyOrChars( stokens[i] );
-            }
-            if ( tokens.length > 5 )
+            String[] tokens = splitAndTokenize( pattern );
+            if ( tokens.length < 1 || tokens.length > 5 )
             {
                 throw new IllegalArgumentException( "Invalid pattern: " + pattern );
             }
-            // we only accept 5 tokens if the classifier = '*'
+
+            ArrayList<Pattern> patterns = new ArrayList<>( 5 );
+
             if ( tokens.length == 5 )
             {
-                if ( tokens[3] != ANY )
+                // trivial, full pattern w/ classifier: G:A:T:C:V
+                patterns.add( toPattern( tokens[0], Coordinate.GROUP_ID ) );
+                patterns.add( toPattern( tokens[1], Coordinate.ARTIFACT_ID ) );
+                patterns.add( toPattern( tokens[2], Coordinate.TYPE ) );
+                patterns.add( toPattern( tokens[3], Coordinate.CLASSIFIER ) );
+                patterns.add( toPattern( tokens[4], Coordinate.BASE_VERSION ) );
+            }
+            else if ( tokens.length == 4 )
+            {
+                // trivial, full pattern w/o classifier: G:A:T:V
+                patterns.add( toPattern( tokens[0], Coordinate.GROUP_ID ) );
+                patterns.add( toPattern( tokens[1], Coordinate.ARTIFACT_ID ) );
+                patterns.add( toPattern( tokens[2], Coordinate.TYPE ) );
+                patterns.add( toPattern( tokens[3], Coordinate.BASE_VERSION ) );
+            }
+            else if ( tokens.length == 3 )
+            {
+                // tricky: may be "*:artifact:*" but also "*:war:*"
+
+                // *:*:* -> ALL
+                // *:*:xxx -> TC(xxx)
+                // *:xxx:* -> AT(xxx)
+                // *:xxx:yyy -> GA(xxx) + TC(XXX)
+                // xxx:*:* -> GA(xxx)
+                // xxx:*:yyy -> G(xxx) + TC(yyy)
+                // xxx:yyy:* -> G(xxx)+A(yyy)
+                // xxx:yyy:zzz -> G(xxx)+A(yyy)+T(zzz)
+                if ( ANY.equals( tokens[0] ) && ANY.equals( tokens[1] ) && ANY.equals( tokens[2] ) )
```

Comment:
I have a problem with those:
* Those aren't documented anywhere
* How can we be sure that those aren't ambigious?

Reply:
This logic just re-implement what happened before (old and older code). My intention is not to fix the pattern spec or change any behavior (as test proves).

Reply:
Alright, agreed. Please put a comment above that those patterns are in question since they aren't documented. An open TODO.

Reply:
pushed 349dd8031f0fed3e9f4e842c7f14da5e1220d52f